### DISCUSSION TITLE
AutoFunctionInvocationFilter - terminating function execution

### DISCUSSION THREAD

**Discussion by deepinderdeol** [microsoft/semantic-kernel#9629] [Q&A] [answered]
Is it possible to terminate function calling once an error is encountered with executing one of the actions specified in a user request?

**Example:** 
- user asks for three actions to be completed in a single request like _"Do Action 1, Action 2, Action 3"_
- Action 1 is executed fine
- An error occurs during executing Action 2

Here we can call **context.Terminate( )** to stop further function execution but this seems to cause issues with chat-history because Assistant message had three tool calls but only one tool message.

Terminating function calling is important because the three actions might have to be executed in that sequence and if Action 2 could not be completed then Action 3 should not be executed. Additionally we want to keep the same chat history.

**Here's the error on issuing a new request after calling context.Terminate( ):**

An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'. The following tool_call_ids did not have response messages: call_3Zh 26yyqUsynQlYFBziYOYLH, call_Gm 4W 0oRWSiPo 9qdutYzt 6F 65
Status: 400 (model_error)
Content:
{
"error": {
"message": "An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'. The following tool_call_ids did not have response messages: call_3Zh 26yyqUsynQlYFBziYOYLH, call_Gm 4W 0oRWSiPo 9qdutYzt 6F 65",
"type": "invalid_request_error",
"param": "messages.[4].role",
"code": null
}
}

**Comment by dmytrostruk** [comment, marked as answer]
Hi @deepinderdeol , thanks for your question!

> Here we can call context.Terminate( ) to stop further function execution but this seems to cause issues with chat-history because Assistant message had three tool calls but only one tool message.

Yes, if you terminate function execution, but you still want to use the same chat history for further requests, you have to manually update chat history with some response for specific `tool_call_id`, like in this example:
https://github.com/microsoft/semantic-kernel/blob/7d9f834130042f4eb50946dbc94667270157b22a/dotnet/samples/Concepts/FunctionCalling/FunctionCalling.cs#L262-L266

The other option would be to use `FunctionChoiceBehavior.None()`. In this case, AI model could suggest which functions to call, so you can invoke them manually, but it won't require a response for each function, which will allow you to terminate the process and still use the same chat history for further requests without modifying it.
https://github.com/microsoft/semantic-kernel/blob/7d9f834130042f4eb50946dbc94667270157b22a/dotnet/samples/Concepts/FunctionCalling/FunctionCalling.cs#L100-L112

I hope this helps. Thank you!

**Reply by deepinderdeol** [reply]
Thanks @dmytrostruk
